File: cpp/FSM.cpp
/*
 * =====================================================================================
 *
 *       Filename:  FSM.cpp
 *
 *    Description:  
 *
 *        Version:  1.0
 *        Created:  11/17/2012 12:03:14 AM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  witwolf (ALL), wo dot bian dot shi dot mo0 at gmail dot com
 *        Company:  ...
 *
 * =====================================================================================
 */

#include "FSM.h"
#include "Utils.h"
#include <queue>
using namespace std;
FSM::FSM()
{
    m_start = new State(NULL,0);
}
FSM::~FSM()
{
    delete m_start;
}
void FSM::init(const vector<string> &words)
{
    State * state;
    vector<string>::const_iterator iter = words.begin();
    for(; iter!=words.end(); iter++)
	{
	    state = m_start;
	    for(int i=0;i<iter->length();i++)
		{
		    state = insertState(state,toLowercase((*iter)[i]));
		}
	    state->m_isEnd = true;
	    state->m_match = *iter;
	}
    
    queue<State *> states;
    states.push(m_start);
    while(!states.empty())
	{
	    state = states.front();
	    states.pop();
	    initFailure(state);
	    for(map<char,State*>::iterator iter = state->m_nexts.begin();
		iter != state->m_nexts.end();
		iter++)
		{
		    states.push(iter->second);
		}
	}
	    
}
State* FSM::nextState(State * curState,char input)
{
    if(curState->m_nexts.count(input))
	{
	    return curState->m_nexts.find(input)->second;
	}
    else if(curState == m_start)
	{
	    return curState;
	}
    else
	{
	    return nextState(curState->m_failureState,input);
	}
}
State * FSM::begin()
{
    return m_start;
}

State * FSM::insertState(State *state,char input)
{
    State *next = NULL ;
    if(state->m_nexts.count(input))
	next = state->m_nexts.find(input)->second;
    else
	{
	    next = new State(state,input);
	    state->m_nexts[input] = next;
	}
    return next ;
}
    
void FSM::initFailure(State * state)
{
    if( !state->m_prev || state->m_prev == m_start )
	{
	    state->m_failureState = m_start ;
	    return;
	}
    else if( state->m_prev->m_failureState->m_nexts.count(state->m_input))
	{
	    state->m_failureState = state->m_prev->m_failureState->m_nexts.find(state->m_input)->second;
	    return;
	}
    else
	state->m_failureState = m_start;
}

State::State(State* prev,char input)
    :m_prev(prev)
    ,m_input(input)
    ,m_failureState(NULL)
    ,m_isEnd(false)
{}

State::~State()
{
    map<char,State*>::iterator iter = m_nexts.begin();
    for(;iter != m_nexts.end(); iter++)
	{
	    delete iter->second;
	}
}

	    

File: java/FSM.java

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.LinkedList;


public class FSM {

	private State start ;
	private State curState;

	public FSM()
	{
		start = new State(null,null ) ;
		curState = start;
	}

	public void init(List<String>  words)
	{
		for(String word:words)
		{
			State state = start ;
			for( int index = 0  ; index < word.length() ; index ++)
			{
				char lowercase = Utils.toLowercase(word.charAt(index)) ;
				state = insertState(state,lowercase);
			}
			state.match = word ;
			state.isEnd = true ;
		}

		Queue<State> states = new LinkedList<State>() ;
		states.add(start) ;
		while(!states.isEmpty())
		{
			State state = states.poll() ;
			initFailureState(state) ;
			for(State s : state.nexts.values() )
			{
				states.add(s) ;
			}	
		}	
	}

	public void nextState(char input)
	{
		Integer key = Integer.valueOf(input);
		if(curState.nexts.containsKey(key))
		{
			curState = curState.nexts.get(key);
			return;
		}
		else if(curState == start)
		{
			return;
		}
		else
		{
			curState = curState.failureState;
			nextState(input);
		}		
	}
	
	public boolean isEnd()
	{
		return curState.isEnd;
	}
	
	public String getMatch()
	{
		return curState.match;
	}

	private State insertState(State state,char input)
	{
		State ret;
		Integer key = Integer.valueOf(input);
		if(state.nexts.containsKey(key))
		{
			ret = state.nexts.get(key);
		}
		else
		{
			ret = new State(state,key);	
			state.nexts.put(key, ret);
		}		
		return ret;
	}
	private void initFailureState(State state)
	{
		if(state.prev == null || state.prev == start)
		{
			state.failureState = start;
		}
		else if(state.prev.failureState.nexts.containsKey(state.input))
		{
			state.failureState = state.prev.failureState.nexts.get(state.input);
		}
		else
		{
			state.failureState = start;
		}
	}
	class State
	{
		public Integer input ; 				
		public State prev ;						
		public Map<Integer,State> nexts;  
		public State failureState ;					
		public boolean isEnd ;						
		public String match ;						

		public State(State _prev,Integer _input)
		{
			nexts = new HashMap<Integer,State>() ;
			prev = _prev ;
			input = _input ;
			failureState = prev ;
			isEnd = false ;
			match = null ;
		}
	} 
}
